Sparkle/SUPlainInstaller.m

```diff
@@ -7,13 +7,165 @@
 //
 
 #import "SUPlainInstaller.h"
-#import "SUPlainInstallerInternals.h"
+#import "SUFileManager.h"
 #import "SUCodeSigningVerifier.h"
 #import "SUConstants.h"
 #import "SUHost.h"
+#import "SULog.h"
 
 @implementation SUPlainInstaller
 
+// Returns the bundle version from the specified host that is appropriate to use as a filename, or nil if we're unable to retrieve one
++ (NSString *)bundleVersionAppropriateForFilenameFromHost:(SUHost *)host
+{
+    NSString *bundleVersion = [host objectForInfoDictionaryKey:(NSString *)kCFBundleVersionKey];
+    NSString *trimmedVersion = @"";
+    
+    if (bundleVersion != nil) {
+        NSMutableCharacterSet *validCharacters = [NSMutableCharacterSet alphanumericCharacterSet];
+        [validCharacters formUnionWithCharacterSet:[NSCharacterSet characterSetWithCharactersInString:@".-()"]];
+        
+        trimmedVersion = [bundleVersion stringByTrimmingCharactersInSet:[validCharacters invertedSet]];
+    }
+    
+    return trimmedVersion.length > 0 ? trimmedVersion : nil;
+}
+
++ (BOOL)performInstallationToURL:(NSURL *)installationURL fromUpdateAtURL:(NSURL *)newURL withHost:(SUHost *)host error:(NSError * __autoreleasing *)error
+{
+    SUFileManager *fileManager = [[SUFileManager alloc] init];
+    
+    // Create a temporary directory for our new app that resides on our destination's volume
+    NSURL *tempNewDirectoryURL = [fileManager makeTemporaryDirectoryWithPreferredName:[installationURL.lastPathComponent.stringByDeletingPathExtension stringByAppendingString:@" (Incomplete Update)"] appropriateForDirectoryURL:installationURL.URLByDeletingLastPathComponent error:error];
```

Comment:
Creating a temp directory for this kind of operation is consistent with how some of the Cocoa APIs work. For example, the docs for replaceItemAtURL:withItemAtURL:backupItemName:options:resultingItemURL:error allows this, but wouldn't allow what you're suggesting. URLForDirectory:inDomain:appropriateForURL:create:error: may even do this as a last ditch effort. They are used for single files (e.g., documents).

Parking the temp directory in the same directory as the final destination is a last resort effort that won't be hit often. Of course we wouldn't want the user to see the incomplete directory visible, but if it's in a temporary location and the installation is interrupted, it may be cleaned out after a couple days or next reboot. Further, there might be some safety implications with having the temp directory be at an unpredictable/hashed location.

I'd rather use a provided Cocoa/System API for creating a temp file/directory where it's viable... I don't think this bit is much different from the old code, except it preferred using the trash instead of a temp location, which is worse.